In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("./datasets/covid19_tweets.csv")

In [3]:
df.sample(5)
# mostra 5 linhas aleatorias

user_name                user_location  \
3642                  Feisty Scotland                          NaN   
15908               Alcohol Awareness            Greater Vancouver   
61498   The Smith OBrien class action  Wellington, New South Wales   
175056           davie ewan macdonald                       Sydney   
154006                   Carrie Grove                  Waltham, MA   

                                         user_description  \
3642    Feminist, anti-Trident and passionate campaign...   
15908   JMF Research BC-Specializing in Alcohol ,Commu...   
61498                                                 NaN   
175056  London/Scot WorkingClass Intellectual Thinker ...   
154006  Award winning print solutions provider. Graphi...   

               user_created  user_followers  user_friends  user_favourites  \
3642    2012-04-30 15:02:25             741           758            26562   
15908   2011-11-05 02:21:51            8813          6578           281320   
61498   2011-07-17 00:01:59             975          2357              521   
175056  2009-04-18 02:18:32            1872          2728             9790   
154006  2009-04-03 15:32:38            1655          1895             1452   

        user_verified                 date  \
3642            False  2020-07-25 10:15:05   
15908           False  2020-07-25 00:40:56   
61498           False  2020-08-02 20:17:46   
175056          False  2020-08-29 23:57:05   
154006          False  2020-08-18 16:15:59   

                                                     text  \
3642    Help slow the spread of #COVID19 and identify ...   
15908   @jjhorgan @MikeSmythNews #alcoholawareness For...   
61498   @SebCostello9 I m interested in why you would ...   
175056  #COVID19 "DOOM AND GLOOM" Yesterday #Insiders ...   
154006  Health and Safety Signage for Schools https://...   

                                                 hashtags              source  \
3642                                          ['COVID19']  Twitter for iPhone   
15908                                ['alcoholawareness']     Twitter Web App   
61498                                         ['covid19']  Twitter for iPhone   
175056                            ['COVID19', 'Insiders']     Twitter Web App   
154006  ['massachusetts', 'publicschools', 'covid19', ...             HubSpot   

        is_retweet  
3642         False  
15908        False  
61498        False  
175056       False  
154006       False

In [4]:
df.loc[df["user_location"].isna()]

user_name user_location  \
7               Derbyshire LPC           NaN   
8            Prathamesh Bendre           NaN   
10      Voice Of CBSE Students           NaN   
16              ChennaiCityNow           NaN   
18                  Dorian Aur           NaN   
...                        ...           ...   
179083           Auntie Kittie           NaN   
179091             🦋  ROGUE  🦋           NaN   
179092       claudia fernandez           NaN   
179097               rainmaker           NaN   
179098               John Geer           NaN   

                                         user_description  \
7                                                     NaN   
8        A poet, reiki practitioner and a student of law.   
10                                                    NaN   
16      Individual tweeting about significant happenin...   
18                                                    NaN   
...                                                   ...   
179083  I'm the auntie you always wished you had.  If ...   
179091  May all aching hearts find their way there thr...   
179092                                                NaN   
179097                                                NaN   
179098  #StayAtHome #StayAtHomeSaveLifes #MaskUp \nFor...   

               user_created  user_followers  user_friends  user_favourites  \
7       2012-02-03 18:08:10             608           355               95   
8       2015-04-25 08:15:41              25            29               18   
10      2020-07-14 17:50:30               8            10                7   
16      2009-04-26 09:38:11            3987            53              749   
18      2011-01-30 18:40:16              46           108              453   
...                     ...             ...           ...              ...   
179083  2019-12-25 18:36:22             888           545               33   
179091  2017-01-30 07:38:23             628           386            14845   
179092  2015-07-02 23:07:03              68           192             2455   
179097  2017-08-16 22:12:17             709          1158            95006   
179098  2020-04-18 01:55:14              61           168            10817   

        user_verified                 date  \
7               False  2020-07-25 12:27:00   
8               False  2020-07-25 12:26:59   
10              False  2020-07-25 12:26:53   
16              False  2020-07-25 12:26:44   
18              False  2020-07-25 12:26:43   
...               ...                  ...   
179083          False  2020-08-29 19:45:02   
179091          False  2020-08-29 19:44:51   
179092          False  2020-08-29 19:44:49   
179097          False  2020-08-29 19:44:35   
179098          False  2020-08-29 19:44:34   

                                                     text  \
7       You now have to wear face coverings when out s...   
8       Praying for good health and recovery of @Chouh...   
10      49K+ Covid19 cases still no response from \n@c...   
16      July 25 #COVID19 update\n#TamilNadu - 6988\nDi...   
18      It is during our darkest moments that we must ...   
...                                                   ...   
179083  Just a quick note for my lovely members.  Soci...   
179091  @dougmillsnyt @realDonaldTrump @FLOTUS "The #T...   
179092  @politvidchannel The Trump Administration's In...   
179097  @politvidchannel Just not the lives of #COVID1...   
179098  Report #COVID19 outbreaks in K-12 schools here...   

                                                 hashtags  \
7                                                     NaN   
8                            ['covid19', 'covidPositive']   
10                                                    NaN   
16                    ['COVID19', 'TamilNadu', 'chennai']   
18                                              ['light']   
...                                                   ...   
179083                                                NaN   


Nesse momento, podemos perceber que o DataFrame possui valores classificados como NaN

algumas operações podem ser afetadas por tais valores, como plot, manipulação de string...

Portanto, no processo de limpeza de dados, devemos sempre analisar se as demais informações dessas colunas são relevantes, ou se podemos simplesmente remover os dados por completo

In [5]:
df.loc[df["user_description"].isna()]

user_name  user_location user_description  \
7               Derbyshire LPC            NaN              NaN   
10      Voice Of CBSE Students            NaN              NaN   
18                  Dorian Aur            NaN              NaN   
20           The Voice of GenX            NaN              NaN   
30                      Ms Paz  United States              NaN   
...                        ...            ...              ...   
178978               Cody Vest            NaN              NaN   
178988                   Shane        Ireland              NaN   
179037            Carol London            NaN              NaN   
179092       claudia fernandez            NaN              NaN   
179097               rainmaker            NaN              NaN   

               user_created  user_followers  user_friends  user_favourites  \
7       2012-02-03 18:08:10             608           355               95   
10      2020-07-14 17:50:30               8            10                7   
18      2011-01-30 18:40:16              46           108              453   
20      2011-05-28 15:28:21             292          1037               58   
30      2019-09-15 18:10:09             127           974            30217   
...                     ...             ...           ...              ...   
178978  2011-04-26 03:13:18             341           170             3555   
178988  2010-09-29 00:07:38            4296          2922            27184   
179037  2009-05-29 16:48:39              96           454           124175   
179092  2015-07-02 23:07:03              68           192             2455   
179097  2017-08-16 22:12:17             709          1158            95006   

        user_verified                 date  \
7               False  2020-07-25 12:27:00   
10              False  2020-07-25 12:26:53   
18              False  2020-07-25 12:26:43   
20              False  2020-07-25 12:26:37   
30              False  2020-07-25 12:26:21   
...               ...                  ...   
178978          False  2020-08-29 19:51:22   
178988          False  2020-08-29 19:50:37   
179037          False  2020-08-29 19:48:10   
179092          False  2020-08-29 19:44:49   
179097          False  2020-08-29 19:44:35   

                                                     text  \
7       You now have to wear face coverings when out s...   
10      49K+ Covid19 cases still no response from \n@c...   
18      It is during our darkest moments that we must ...   
20      @EvanAKilgore @realDonaldTrump Good Patriots!\...   
30      #FEMA acknowledges #PuertoRico lacks rebuilt h...   
...                                                   ...   
178978  My first week back in school has taught me two...   
178988  7 consecutive days without a death from #Covid...   
179037  Trump Program to Cover Uninsured Covid-19 Pati...   
179092  @politvidchannel The Trump Administration's In...   
179097  @politvidchannel Just not the lives of #COVID1...   

                                        hashtags               source  \
7                                            NaN            TweetDeck   
10                                           NaN      Twitter Web App   
18                                     ['light']      Twitter Web App   
20                                           NaN   Twitter for iPhone   
30             ['FEMA', 'PuertoRico', 'COVID19']   Twitter for iPhone   
...                                          ...                  ...   
178978                                       NaN   Twitter for iPhone   
178988                               ['Covid19']  Twitter for Android   
179037   ['HealthInsurance', 'TrumpIsKillingUs']     Twitter for iPad   
179092                                 ['COVID']      Twitter Web App   
179097  ['COVID19', 'coronavirus', 'TrumpVirus']  Twitter for Android   

        is_retweet  
7            False  
10           False  
18           False  
20           False  
30           False  
..

In [6]:
df.shape

(179108, 13)

o DataFrame possui no total 179108 linhas, se excluirmos todos os dados NaN, de ambas as colunas, estariamos excluindo um maximo de 47059 (desconsiderando que algumas linhas possam ter NaN em ambas colunas).

# Excluindo dados

In [7]:
df_exclui = df.dropna(subset=["user_description", "user_location"])

In [8]:
df_exclui.shape

(138352, 13)

In [9]:
df_exclui.loc[df_exclui["user_description"].isna()]

Empty DataFrame
Columns: [user_name, user_location, user_description, user_created, user_followers, user_friends, user_favourites, user_verified, date, text, hashtags, source, is_retweet]
Index: []

In [10]:
df_exclui.loc[df_exclui["user_location"].isna()]

Empty DataFrame
Columns: [user_name, user_location, user_description, user_created, user_followers, user_friends, user_favourites, user_verified, date, text, hashtags, source, is_retweet]
Index: []

# Filtrando dados com valor

In [11]:
df["user_location"] = df["user_location"].fillna("")

In [12]:
df

user_name               user_location  \
0                       ᏉᎥ☻լꂅϮ                  astroworld   
1                Tom Basile 🇺🇸                New York, NY   
2              Time4fisticuffs            Pewee Valley, KY   
3                  ethel mertz        Stuck in the Middle    
4                     DIPR-J&K           Jammu and Kashmir   
...                        ...                         ...   
179103  AJIMATI AbdulRahman O.             Ilorin, Nigeria   
179104                   Jason                     Ontario   
179105             BEEHEMOTH ⏳                   🇨🇦 Canada   
179106           Gary DelPonte               New York City   
179107                 TUKY II  Aliwal North, South Africa   

                                         user_description  \
0       wednesday addams as a disney princess keepin i...   
1       Husband, Father, Columnist & Commentator. Auth...   
2       #Christian #Catholic #Conservative #Reagan #Re...   
3       #Browns #Indians #ClevelandProud #[]_[] #Cavs ...   
4       🖊️Official Twitter handle of Department of Inf...   
...                                                   ...   
179103    Animal Scientist|| Muslim|| Real Madrid/Chelsea   
179104  When your cat has more baking soda than Ninja ...   
179105  ⚒️ The Architects of Free Trade ⚒️ Really Did ...   
179106  Global UX UI Visual Designer. StoryTeller, Mus...   
179107  TOKELO SEKHOPA | TUKY II | LAST BORN | EISH TU...   

               user_created  user_followers  user_friends  user_favourites  \
0       2017-05-26 05:46:42             624           950            18775   
1       2009-04-16 20:06:23            2253          1677               24   
2       2009-02-28 18:57:41            9275          9525             7254   
3       2019-03-07 01:45:06             197           987             1488   
4       2017-02-12 06:45:15          101009           168              101   
...                     ...             ...           ...              ...   
179103  2013-12-30 18:59:19             412          1609             1062   
179104  2011-12-21 04:41:30             150           182             7295   
179105  2016-07-13 17:21:59            1623          2160            98000   
179106  2009-10-27 17:43:13            1338          1111                0   
179107  2018-04-14 17:30:07              97          1697              566   

        user_verified                 date  \
0               False  2020-07-25 12:27:21   
1                True  2020-07-25 12:27:17   
2               False  2020-07-25 12:27:14   
3               False  2020-07-25 12:27:10   
4               False  2020-07-25 12:27:08   
...               ...                  ...   
179103          False  2020-08-29 19:44:21   
179104          False  2020-08-29 19:44:16   
179105          False  2020-08-29 19:44:15   
179106          False  2020-08-29 19:44:14   
179107          False  2020-08-29 19:44:08   

                                                     text  \
0       If I smelled the scent of hand sanitizers toda...   
1       Hey @Yankees @YankeesPR and @MLB - wouldn't it...   
2       @diane3443 @wdunlap @realDonaldTrump Trump nev...   
3       @brookbanktv The one gift #COVID19 has give me...   
4       25 July : Media Bulletin on Novel #CoronaVirus...   
...                                                   ...   
179103  Thanks @IamOhmai for nominating me for the @WH...   
179104  2020! The year of insanity! Lol! #COVID19 http...   
179105  @CTVNews A powerful painting by Juan Lucena. I...   
179106  More than 1,200 students test positive for #CO...   
179107  I stop when I see a Stop\n\n@SABCNews\n@Izinda...   

                                 hashtags               source  is_retweet  
0                                     NaN   Twitter for iPhone       False  
1                                     NaN  Twitter for Android       False  
2                             ['COVID19']  Twitter for Android       False  
3           

In [13]:
df.loc[df["user_location"].isna()]

Empty DataFrame
Columns: [user_name, user_location, user_description, user_created, user_followers, user_friends, user_favourites, user_verified, date, text, hashtags, source, is_retweet]
Index: []

In [14]:
df.loc[df["user_location"] == ""]

user_name user_location  \
7               Derbyshire LPC                 
8            Prathamesh Bendre                 
10      Voice Of CBSE Students                 
16              ChennaiCityNow                 
18                  Dorian Aur                 
...                        ...           ...   
179083           Auntie Kittie                 
179091             🦋  ROGUE  🦋                 
179092       claudia fernandez                 
179097               rainmaker                 
179098               John Geer                 

                                         user_description  \
7                                                     NaN   
8        A poet, reiki practitioner and a student of law.   
10                                                    NaN   
16      Individual tweeting about significant happenin...   
18                                                    NaN   
...                                                   ...   
179083  I'm the auntie you always wished you had.  If ...   
179091  May all aching hearts find their way there thr...   
179092                                                NaN   
179097                                                NaN   
179098  #StayAtHome #StayAtHomeSaveLifes #MaskUp \nFor...   

               user_created  user_followers  user_friends  user_favourites  \
7       2012-02-03 18:08:10             608           355               95   
8       2015-04-25 08:15:41              25            29               18   
10      2020-07-14 17:50:30               8            10                7   
16      2009-04-26 09:38:11            3987            53              749   
18      2011-01-30 18:40:16              46           108              453   
...                     ...             ...           ...              ...   
179083  2019-12-25 18:36:22             888           545               33   
179091  2017-01-30 07:38:23             628           386            14845   
179092  2015-07-02 23:07:03              68           192             2455   
179097  2017-08-16 22:12:17             709          1158            95006   
179098  2020-04-18 01:55:14              61           168            10817   

        user_verified                 date  \
7               False  2020-07-25 12:27:00   
8               False  2020-07-25 12:26:59   
10              False  2020-07-25 12:26:53   
16              False  2020-07-25 12:26:44   
18              False  2020-07-25 12:26:43   
...               ...                  ...   
179083          False  2020-08-29 19:45:02   
179091          False  2020-08-29 19:44:51   
179092          False  2020-08-29 19:44:49   
179097          False  2020-08-29 19:44:35   
179098          False  2020-08-29 19:44:34   

                                                     text  \
7       You now have to wear face coverings when out s...   
8       Praying for good health and recovery of @Chouh...   
10      49K+ Covid19 cases still no response from \n@c...   
16      July 25 #COVID19 update\n#TamilNadu - 6988\nDi...   
18      It is during our darkest moments that we must ...   
...                                                   ...   
179083  Just a quick note for my lovely members.  Soci...   
179091  @dougmillsnyt @realDonaldTrump @FLOTUS "The #T...   
179092  @politvidchannel The Trump Administration's In...   
179097  @politvidchannel Just not the lives of #COVID1...   
179098  Report #COVID19 outbreaks in K-12 schools here...   

                                                 hashtags  \
7                                                     NaN   
8                            ['covid19', 'covidPositive']   
10                                                    NaN   
16                    ['COVID19', 'TamilNadu', 'chennai']   
18                                              ['light']   
...                                                   ...   
179083                                                NaN   


In [15]:
df.loc[df["user_description"].isna()]

user_name  user_location user_description  \
7               Derbyshire LPC                             NaN   
10      Voice Of CBSE Students                             NaN   
18                  Dorian Aur                             NaN   
20           The Voice of GenX                             NaN   
30                      Ms Paz  United States              NaN   
...                        ...            ...              ...   
178978               Cody Vest                             NaN   
178988                   Shane        Ireland              NaN   
179037            Carol London                             NaN   
179092       claudia fernandez                             NaN   
179097               rainmaker                             NaN   

               user_created  user_followers  user_friends  user_favourites  \
7       2012-02-03 18:08:10             608           355               95   
10      2020-07-14 17:50:30               8            10                7   
18      2011-01-30 18:40:16              46           108              453   
20      2011-05-28 15:28:21             292          1037               58   
30      2019-09-15 18:10:09             127           974            30217   
...                     ...             ...           ...              ...   
178978  2011-04-26 03:13:18             341           170             3555   
178988  2010-09-29 00:07:38            4296          2922            27184   
179037  2009-05-29 16:48:39              96           454           124175   
179092  2015-07-02 23:07:03              68           192             2455   
179097  2017-08-16 22:12:17             709          1158            95006   

        user_verified                 date  \
7               False  2020-07-25 12:27:00   
10              False  2020-07-25 12:26:53   
18              False  2020-07-25 12:26:43   
20              False  2020-07-25 12:26:37   
30              False  2020-07-25 12:26:21   
...               ...                  ...   
178978          False  2020-08-29 19:51:22   
178988          False  2020-08-29 19:50:37   
179037          False  2020-08-29 19:48:10   
179092          False  2020-08-29 19:44:49   
179097          False  2020-08-29 19:44:35   

                                                     text  \
7       You now have to wear face coverings when out s...   
10      49K+ Covid19 cases still no response from \n@c...   
18      It is during our darkest moments that we must ...   
20      @EvanAKilgore @realDonaldTrump Good Patriots!\...   
30      #FEMA acknowledges #PuertoRico lacks rebuilt h...   
...                                                   ...   
178978  My first week back in school has taught me two...   
178988  7 consecutive days without a death from #Covid...   
179037  Trump Program to Cover Uninsured Covid-19 Pati...   
179092  @politvidchannel The Trump Administration's In...   
179097  @politvidchannel Just not the lives of #COVID1...   

                                        hashtags               source  \
7                                            NaN            TweetDeck   
10                                           NaN      Twitter Web App   
18                                     ['light']      Twitter Web App   
20                                           NaN   Twitter for iPhone   
30             ['FEMA', 'PuertoRico', 'COVID19']   Twitter for iPhone   
...                                          ...                  ...   
178978                                       NaN   Twitter for iPhone   
178988                               ['Covid19']  Twitter for Android   
179037   ['HealthInsurance', 'TrumpIsKillingUs']     Twitter for iPad   
179092                                 ['COVID']      Twitter Web App   
179097  ['COVID19', 'coronavirus', 'TrumpVirus']  Twitter for Android   

        is_retweet  
7            False  
10           False  
18           False  
20           False  
30           False  
..

In [16]:
# podemos usar qualquer valor como filtro

df["user_description"] = df["user_description"].fillna("minha_desc")

In [17]:
df.loc[df["user_description"].str.contains("minha_desc")]

user_name  user_location user_description  \
7               Derbyshire LPC                      minha_desc   
10      Voice Of CBSE Students                      minha_desc   
18                  Dorian Aur                      minha_desc   
20           The Voice of GenX                      minha_desc   
30                      Ms Paz  United States       minha_desc   
...                        ...            ...              ...   
178978               Cody Vest                      minha_desc   
178988                   Shane        Ireland       minha_desc   
179037            Carol London                      minha_desc   
179092       claudia fernandez                      minha_desc   
179097               rainmaker                      minha_desc   

               user_created  user_followers  user_friends  user_favourites  \
7       2012-02-03 18:08:10             608           355               95   
10      2020-07-14 17:50:30               8            10                7   
18      2011-01-30 18:40:16              46           108              453   
20      2011-05-28 15:28:21             292          1037               58   
30      2019-09-15 18:10:09             127           974            30217   
...                     ...             ...           ...              ...   
178978  2011-04-26 03:13:18             341           170             3555   
178988  2010-09-29 00:07:38            4296          2922            27184   
179037  2009-05-29 16:48:39              96           454           124175   
179092  2015-07-02 23:07:03              68           192             2455   
179097  2017-08-16 22:12:17             709          1158            95006   

        user_verified                 date  \
7               False  2020-07-25 12:27:00   
10              False  2020-07-25 12:26:53   
18              False  2020-07-25 12:26:43   
20              False  2020-07-25 12:26:37   
30              False  2020-07-25 12:26:21   
...               ...                  ...   
178978          False  2020-08-29 19:51:22   
178988          False  2020-08-29 19:50:37   
179037          False  2020-08-29 19:48:10   
179092          False  2020-08-29 19:44:49   
179097          False  2020-08-29 19:44:35   

                                                     text  \
7       You now have to wear face coverings when out s...   
10      49K+ Covid19 cases still no response from \n@c...   
18      It is during our darkest moments that we must ...   
20      @EvanAKilgore @realDonaldTrump Good Patriots!\...   
30      #FEMA acknowledges #PuertoRico lacks rebuilt h...   
...                                                   ...   
178978  My first week back in school has taught me two...   
178988  7 consecutive days without a death from #Covid...   
179037  Trump Program to Cover Uninsured Covid-19 Pati...   
179092  @politvidchannel The Trump Administration's In...   
179097  @politvidchannel Just not the lives of #COVID1...   

                                        hashtags               source  \
7                                            NaN            TweetDeck   
10                                           NaN      Twitter Web App   
18                                     ['light']      Twitter Web App   
20                                           NaN   Twitter for iPhone   
30             ['FEMA', 'PuertoRico', 'COVID19']   Twitter for iPhone   
...                                          ...                  ...   
178978                                       NaN   Twitter for iPhone   
178988                               ['Covid19']  Twitter for Android   
179037   ['HealthInsurance', 'TrumpIsKillingUs']     Twitter for iPad   
179092                                 ['COVID']      Twitter Web App   
179097  ['COVID19', 'coronavirus', 'TrumpVirus']  Twitter for Android   

        is_retweet  
7            False  
10           False  
18           False  
20           False  
30           False  
..